watcher.go

```diff
@@ -409,101 +356,256 @@ func (w *Watcher) Stop() {
+func (t *tracker) viewCount() int {
+	t.Lock()
+	defer t.Unlock()
+	return len(t.views)
+}
+
+// notUsed clears the inUse flag, for testing
+func (t *tracker) notUsed(notifierID, viewID string) bool {
+	for i, tp := range t.tracked {
+		if tp.view == viewID && tp.notify == notifierID {
+			t.tracked[i] = tp.refresh()
+			return true
+		}
 	}
+	return false
 }
 
-func (t *tracker) update(tmplID string, deps ...dep.Dependency) {
-	t.clear(tmplID)
+// lookup returns the view and true, or view's zero value and false
+func (t *tracker) lookup(viewID string) (*view, bool) {
 	t.Lock()
 	defer t.Unlock()
-	for _, d := range deps {
-		t.deps.add(d.String(), tmplID)
-		t.tpls.add(tmplID, d.String())
+	v, ok := t.views[viewID]
+	return v, ok
+}
+
+// Adds new entry
+func (t *tracker) add(v *view, n Notifier) {
+	t.Lock()
+	defer t.Unlock()
+	t.views[v.ID()] = v
+	t.notifiers[n.ID()] = n
+	t.tracked = append(t.tracked,
+		trackedPair{view: v.ID(), notify: n.ID(), inUse: true})
+}
+
+// Marks all trackedPairs w/ a view as having been used
+func (t *tracker) usedID(viewID string) {
+	t.Lock()
+	defer t.Unlock()
+	for idx, tp := range t.tracked {
+		if tp.view == viewID {
+			t.tracked[idx] = tp.used()
+		}
+	}
+}
+
+// Remove view and all trackedPairs that contained it
+func (t *tracker) remove(viewID string) bool {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.views, viewID)
+	tmp := t.tracked[:0]
+	var removed bool
+	for _, tp := range t.tracked {
+		if tp.view != viewID {
+			tmp = append(tmp, tp)
+		} else {
+			removed = true
+		}
 	}
+	t.tracked = tmp
+	return removed
 }
 
-func (t *tracker) clear(tmplID string) {
+// stop all view from polling/watching
+func (t *tracker) stopViews() {
 	t.Lock()
 	defer t.Unlock()
-	for d := range t.tpls[tmplID] {
-		delete(t.deps[d], tmplID)
-		delete(t.tpls[tmplID], d)
+	for id, view := range t.views {
+		delete(t.views, id)
+		if view == nil {
+			continue
+		}
+		view.stop()
+	}
+}
+
+// Return all views for a notifier
+func (t *tracker) notifiersFor(v *view) []Notifier {
+	viewID := v.ID()
+	results := make([]Notifier, 0, 8)
+	for _, tp := range t.tracked {
+		if tp.view == viewID {
+			results = append(results, t.notifiers[tp.notify])
+		}
 	}
+	return results
+}
+
+// initialized returns true if the view has had its data fetched at least once
+func (t *tracker) initialized(viewID string) bool {
+	return t.views[viewID].receivedData
 }
 
-func (t *tracker) findUnused(depIDs []string) map[string]struct{} {
-	t.RLock()
-	defer t.RUnlock()
-	result := make(map[string]struct{})
-	for _, depID := range depIDs {
-		if len(t.deps[depID]) == 0 {
-			result[depID] = struct{}{}
+// complete returns true if every dependency used has been initialized
+// ie. it returns true if all values have been fetched
+func (t *tracker) complete(n Notifier) bool {
+	for _, tp := range t.tracked {
+		thisNotifier := tp.notify == n.ID()
+		//		fmt.Println(thisNotifier, tp.inUse, !t.initialized(tp.view))
+		//		fmt.Printf("%#v\n", tp)
```

Comment:
Cleanup!

Reply:
I had left these here on purpose as I kept re-adding them while debugging... but I'll be optimistic and remove them for now.